piccolo/columns/m2m.py

```diff
@@ -0,0 +1,336 @@
+                "{m2m_table_name}"."{fk_1_name}" = "inner_{table_1_name}"."{table_1_pk_name}"
+            )
+            JOIN "{table_2_name}" "inner_{table_2_name}" ON (
+                "{m2m_table_name}"."{fk_2_name}" = "inner_{table_2_name}"."{table_2_pk_name}"
+            )
+            WHERE "{m2m_table_name}"."{fk_1_name}" = "{table_1_name}"."{table_1_pk_name}"
+        """  # noqa: E501
+
+        if engine_type == "postgres":
+            return f"""
+                ARRAY(
+                    SELECT
+                        "inner_{table_2_name}"."{column_name}"
+                    FROM {inner_select}
+                ) AS "{m2m_relationship_name}"
+            """
+        elif engine_type == "sqlite":
+            return f"""
+                (
+                    SELECT group_concat(
+                        "inner_{table_2_name}"."{column_name}"
+                    )
+                    FROM {inner_select}
+                )
+                AS "{m2m_relationship_name} [M2M]"
+            """
+        else:
+            raise ValueError(f"{engine_type} is an unrecognised engine type")
+
+
+@dataclass
+class M2MMeta:
+    joining_table: t.Union[t.Type[Table], LazyTableReference]
+    _foreign_key_columns: t.Optional[t.List[ForeignKey]] = None
+
+    # Set by the Table Metaclass:
+    _name: t.Optional[str] = None
+    _table: t.Optional[t.Type[Table]] = None
+
+    @property
+    def name(self) -> str:
+        if not self._name:
+            raise ValueError(
+                "`_name` isn't defined - the Table Metaclass should set it."
+            )
+        return self._name
+
+    @property
+    def table(self) -> t.Type[Table]:
+        if not self._table:
+            raise ValueError(
+                "`_table` isn't defined - the Table Metaclass should set it."
+            )
+        return self._table
+
+    @property
+    def resolved_joining_table(self) -> t.Type[Table]:
+        """
+        Evaluates the ``joining_table`` attribute if it's a
+        ``LazyTableReference``, raising a ``ValueError`` if it fails, otherwise
+        returns a ``Table`` subclass.
+        """
+        from piccolo.table import Table
+
+        if isinstance(self.joining_table, LazyTableReference):
+            return self.joining_table.resolve()
+        elif inspect.isclass(self.joining_table) and issubclass(
+            self.joining_table, Table
+        ):
+            return self.joining_table
+        else:
+            raise ValueError(
+                "The joining_table attribute is neither a Table subclass or a "
+                "LazyTableReference instance."
+            )
+
+    @property
+    def foreign_key_columns(self) -> t.List[ForeignKey]:
+        if not self._foreign_key_columns:
+            self._foreign_key_columns = (
+                self.resolved_joining_table._meta.foreign_key_columns[0:2]
+            )
+        return self._foreign_key_columns
```

Comment:
```suggestion
            self._foreign_key_columns = (
                self.resolved_joining_table._meta.foreign_key_columns[:2]
            )
        return self._foreign_key_columns
```

The default starting value for a slice is 0, so it is unnecessary to explicitly define it.